### PR TITLE
Implement `Hyperband` pruner

### PR DESCRIPTION
**Pull request by crcrpar**
**UPDATE**
The original version is separated into #805 and this PR.

---

This PR is based on #301.

Intuitively, Hyperband (HB) eliminates the dependency on parameters of SuccessiveHalving (SH) by internally executes multiple SHs with different configurations.

## Design
`Study` with `HyperbandPruner` runs in the same way as with other pruners. `HyperbandPruner` class maintains some number of `SuccessiveHalvingPruner`s (= brackets) and selects a pruner for each `Trial`. So, the algorithm would be different from the paper to some extent. There're two challenges, 1) different trials of the same `Study` have to be pruned by different brackets, and 2) When sampling for a new trial, `Sampler` can only use the trials of the same bracket.

## Major Changes
- `Study` collects the list of trials (= `friend_trials` in the code) and set it as a `Trial`'s attribute
    - To filter trials with some metadata, study sets the information of pruner as `user_attr` to `Trial` and uses it as a filter. 
- `Trial` passes its `friend_trial` to `study.sampler`
- `Sampler`'s sampling methods accept the list of trials as their argument

An alternative design, a new class that manages multiple `Study`s is implemented in https://github.com/crcrpar/optuna/tree/dev/study-manager.

### CODE REVIEW

File: optuna/trial.py [hunk 1]
```diff
@@ -136,6 +138,24 @@ def __init__(
...
+    def _clear_friend_trials(self):
+        # type: () -> None
+
+        self._friend_trials = None
+
+    @property
+    def friend_trials(self):
+        # type: () -> Optional[List[FrozenTrial]]
+
+        return self._friend_trials
```
Comment (crcrpar) [inline optuna/trial.py]: I don't think the naming of `friend_trials` is cool 😅

**Comment by crcrpar** [comment]
As I have done in general and I want to get any feedback, so I mark this as ready for review.

File: optuna/pruners/base.py [hunk 2]
```diff
@@ -30,3 +31,26 @@ def prune(self, study, trial):
         """
 
         raise NotImplementedError
+
+    @abc.abstractmethod
+    def get_trial_pruner_auxiliary_data(self, study_id, trial_number):
```
Comment (crcrpar) [inline optuna/pruners/base.py]: As `study_id` is deprecated, should this be `study_name` as follows?
```suggestion
    def get_trial_pruner_auxiliary_data(self, study_name, trial_number):
```

File: optuna/pruners/base.py [hunk 3]
```diff
@@ -30,3 +31,26 @@ def prune(self, study, trial):
...
+        # type: (int, int) -> str
```
Comment (crcrpar) [inline optuna/pruners/base.py]: Reflecting the above change.
```suggestion
        # type: (str, int) -> str
```

File: optuna/pruners/hyperband.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+    def _calc_bracket_resource_budget(self, pruner_index, n_pruners):
+        # type: (int, int) -> int
+        n = self._reduction_factor ** (n_pruners - 1)
+        budget = n
+        for i in range(pruner_index, n_pruners - 1):
+            budget += n / 2
+        return budget
+
+    def get_bracket_id(self, study_id, trial_number):
```
Comment (crcrpar) [inline optuna/pruners/hyperband.py]: As mentioned above, `study_id` is deprecated.
```suggestion
    def get_bracket_id(self, study_name, trial_number):
```

File: optuna/pruners/hyperband.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        # type: (int, int) -> int
```
Comment (crcrpar) [inline optuna/pruners/hyperband.py]: ditto
```suggestion
        # type: (str, int) -> int
```

File: optuna/pruners/hyperband.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        """Computes the id of bracket for a trial of `trial_number`."""
+
+        n = hash('{}_{}'.format(study_id, trial_number)) % self._resource_badget
```
Comment (crcrpar) [inline optuna/pruners/hyperband.py]: ditto
```suggestion
        n = hash('{}_{}'.format(study_name, trial_number)) % self._resource_badget
```

File: optuna/pruners/nop.py [hunk 7]
```diff
@@ -28,3 +28,13 @@ def prune(self, study, trial):
         """Please consult the documentation for :func:`BasePruner.prune`."""
 
         return False
+
+    def get_trial_pruner_auxiliary_data(self, study_id, trial_number):
```
Comment (crcrpar) [inline optuna/pruners/nop.py]:
```suggestion
    def get_trial_pruner_auxiliary_data(self, study_name, trial_number):
```

File: optuna/pruners/nop.py [hunk 8]
```diff
@@ -28,3 +28,13 @@ def prune(self, study, trial):
...
+        # type: (int, int) -> str
```
Comment (crcrpar) [inline optuna/pruners/nop.py]:
```suggestion
        # type: (str, int) -> str
```

File: optuna/pruners/percentile.py [hunk 9]
```diff
@@ -151,3 +151,13 @@ def prune(self, study, trial):
         if direction == structs.StudyDirection.MAXIMIZE:
             return best_intermediate_result < p
         return best_intermediate_result > p
+
+    def get_trial_pruner_auxiliary_data(self, study_id, trial_number):
```
Comment (crcrpar) [inline optuna/pruners/percentile.py]:
```suggestion
    def get_trial_pruner_auxiliary_data(self, study_name, trial_number):
```

File: optuna/pruners/percentile.py [hunk 10]
```diff
@@ -151,3 +151,13 @@ def prune(self, study, trial):
...
+        # type: (int, int) -> str
```
Comment (crcrpar) [inline optuna/pruners/percentile.py]:
```suggestion
        # type: (str, int) -> str
```

File: optuna/pruners/successive_halving.py [hunk 11]
```diff
@@ -117,17 +127,27 @@ def prune(self, study, trial):
...
-            study._storage.set_trial_system_attr(trial._trial_id, _completed_rung_key(rung), value)
+            study._storage.set_trial_system_attr(
+                trial._trial_id, self._completed_rung_key(rung), value)
             direction = study.direction
             if not self._is_promotable(rung, value, all_trials, direction):
                 return True
 
             rung += 1
 
+    def get_trial_pruner_auxiliary_data(self, study_id, trial_number):
```
Comment (crcrpar) [inline optuna/pruners/successive_halving.py]:
```suggestion
    def get_trial_pruner_auxiliary_data(self, study_name, trial_number):
```

File: optuna/pruners/successive_halving.py [hunk 12]
```diff
@@ -117,17 +127,27 @@ def prune(self, study, trial):
...
+        # type: (int, int) -> str
```
Comment (crcrpar) [inline optuna/pruners/successive_halving.py]:
```suggestion
        # type: (str, int) -> str
```

File: optuna/testing/integration.py [hunk 13]
```diff
@@ -12,6 +12,16 @@ def prune(self, study, trial):
 
         return self.is_pruning
 
+    def get_trial_pruner_auxiliary_data(self, study_id, trial_number):
```
Comment (crcrpar) [inline optuna/testing/integration.py]:
```suggestion
    def get_trial_pruner_auxiliary_data(self, study_name, trial_number):
```

File: optuna/testing/integration.py [hunk 14]
```diff
@@ -12,6 +12,16 @@ def prune(self, study, trial):
...
+        # type: (int, int) -> str
```
Comment (crcrpar) [inline optuna/testing/integration.py]:
```suggestion
        # type: (str, int) -> str
```

File: optuna/pruners/hyperband.py [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+
+    def get_bracket_id(self, study_id, trial_number):
+        # type: (int, int) -> int
+        """Computes the index of bracket for a trial of ``trial_number``.
+
+        The index of a bracket is noted as :math:`s` in
+        `Hyperband paper <http://www.jmlr.org/papers/volume18/16-558/16-558.pdf>`_.
+        """
+
+        n = hash('{}_{}'.format(study_id, trial_number)) % self._resource_badget
```
Comment (c-bata) [inline optuna/pruners/hyperband.py]: I'm not confident but It looks `trial_number % self._resource_budget` is enough, right?

```suggestion
        n =  trial_number % self._resource_budget
```

File: optuna/pruners/hyperband.py [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+            self,
+            min_resource=1,
+            reduction_factor=3,
+            min_early_stopping_rate_low=0,
+            min_early_stopping_rate_high=4
+    ):
+        # type: (int, int, int, int) -> None
+        self._pruners = []  # type: List[SuccessiveHalvingPruner]
+        self._reduction_factor = reduction_factor
+        self._resource_badget = 0
```
Comment (c-bata) [inline optuna/pruners/hyperband.py]: ditto.

```suggestion
        self._resource_budget = 0
```

File: optuna/pruners/hyperband.py [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+        n_pruners = min_early_stopping_rate_high - min_early_stopping_rate_low + 1
+        self._n_pruners = n_pruners
+        self._bracket_resource_budgets = []  # type: List[int]
+
+        _logger.debug('Hypberband has {} brackets'.format(self.n_pruners))
+
+        for i in range(n_pruners):
+            bracket_resource_budget = self._calc_bracket_resource_budget(i, n_pruners)
+            self._resource_badget += bracket_resource_budget
```
Comment (c-bata) [inline optuna/pruners/hyperband.py]: ditto.

```suggestion
            self._resource_budget += bracket_resource_budget
```

File: optuna/pruners/hyperband.py [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+    @property
+    def resource_budget(self):
+        # type: () -> int
+        """The total amount of resource.
+
+        This value is a proxy for :math:`B` in the
+        `Hyperband paper <http://www.jmlr.org/papers/volume18/16-558/16-558.pdf>`_.
+        """
+
+        return self._resource_badget
```
Comment (c-bata) [inline optuna/pruners/hyperband.py]:
```suggestion
        return self._resource_budget
```

File: optuna/pruners/hyperband.py [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+    ):
+        # type: (int, int, int, int) -> None
+        self._pruners = []  # type: List[SuccessiveHalvingPruner]
+        self._reduction_factor = reduction_factor
+        self._resource_badget = 0
+        n_pruners = min_early_stopping_rate_high - min_early_stopping_rate_low + 1
+        self._n_pruners = n_pruners
+        self._bracket_resource_budgets = []  # type: List[int]
+
+        _logger.debug('Hypberband has {} brackets'.format(self.n_pruners))
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply optuna/pruners/hyperband.py]: Thank you.

**Comment by crcrpar** [comment]
@c-bata 

Thank you for your review! 
Being not complete is not a problem.
I really appreciate your response. 😀

> why this PR includes the change of sampler interface, and what friend_trials means

I think your question is equivalent to why new argument `trials: List[FrozenTrial]`.
A `sampler`, especially TPE sampler, must only reflect the trials that have the same `SuccessiveHalvingPruner` of `HyperbandPruner`  as a `trial` that is currently being initialized.
To realize this, I added an attribute of `pruner_metadata` to `Trial` via `set_uesr_atttr` inside `Study`. Also, make `Study` collects appropriate trials and set them as `friend_trials` to the `trial` tentatively for the ease of trial selection.

I hope this helps you.

**Comment by c-bata** [comment]
Thank you! Probably, I understand.

1. When using SuccessiveHalvingPruner, it has no problem to use the last intermediate score of pruned trials.
2. But it seems that HyperbandPruner is not.
3. So you labeled trials as `pruned_metadata` (this is a string representation of bracket_id in HyperbandPruner) by `get_trial_pruner_auxiliary_data()` method.
4. `friend_trials` returns the trials which has the same `pruned_metadata` (`bracket_id`).
5. So you passes `friend_trials` to samplers.

File: optuna/pruners/hyperband.py (hunk 16 above)
Comment (crcrpar) [reply optuna/pruners/hyperband.py]: thank you!

File: optuna/testing/sampler.py [hunk 20]
```diff
@@ -47,16 +56,22 @@ def sample_independent(self, study, trial, param_name, param_distribution):
...
+            trial,  # type: FrozenTrial
+            search_space,  # type: Dict[str, BaseDistribution]
+            trials=None  # type: Optional[List[FrozenTrial]]
+    ):
+        # type: (...) -> Dict[str, float]
 
         if len(study.trials) > 1:
             raise RuntimeError("`FirstTrialOnlyRandomSampler` only works on the first trial.")
 
         return super(FirstTrialOnlyRandomSampler, self).sample_relative(study, trial, search_space)
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply optuna/testing/sampler.py]: Good catch, thank you!

**Comment by crcrpar** [comment]
This PR consists of two major changes
- new argument of trials to samplers sample methods
- hyperband

Both changes are not trivial I think, thus I'd like to separate this into two PRs.

**Comment by sile** [comment]
> Both changes are not trivial I think, thus I'd like to separate this into two PRs.

Nice idea!

**Comment by sile** [comment]
I think that this PR was taken over by #809. Could we close this? > @crcrpar

**Comment by crcrpar** [comment]
> I think that this PR was taken over by #809. Could we close this? > @crcrpar

of course.

**Pull request closed without merging**